Introduce module `preprocessing` for distributed scaling/normalization

mrfh92 opened the pull request:
Inteded to resolve #400: in this PR we want to implement a preprocessing module for heat that has roughly the same capabilities as [the one of scikit-learn](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html). 

So far I have implemented: 

* StandardScaler 
* MinMaxScaler
* MaxAbsScaler
* RobustScaler
* Normalizer 

... 

**Idea:** use `heat.BaseEstimator` and keep as much of the scikit-learn API as possible

mrfh92 commented:
current problem in `StandardScaler`: sth is wrong with the variance computation...

mrfh92 commented:
@krajsek From your experience in the neuroscience community, is there another preprocessing tool (e.g., from the list of scikit-learn) we should definitively add?

krajsek commented:
@mrfh92 No, not at that level. There are a large number of preprocessing steps used in neuroimaging pipelines but they are more own image processing algorithms like motion correction, slice timing corrction or registration. But on this level, simple normalization etc. I cannot think about any neuroscience specific preprocesssing steps. Apart from that I think OneHotEncoder is quite usefull.

mrfh92 commented:
to be discussed: 
* which additional preprocessing routines (e.g., from those available in `scikit-learn`) could/should be added?
* which of the routines from this module should be part of continuous benchmarking

mrfh92 commented:
extensive citation of scikit-learn should be added in the docs

mrfh92 commented:
@mtar I have addressed most issues you mentioned in the last PR talk 
only type hints for returned `self` seem to be a bit tricky, because `from typing import Self` produces an error for me...

mtar commented:
@mrfh92 the _Self_ type was introduced in Python 3.11, on older versions you can import it from the _typing_extensions_ module

mrfh92 commented:
@mtar thanks for the hint ... i have added the respective changes

ClaudiaComito reviewed (changes requested):
Thank you so much @mrfh92 , this is very timely.  I have a few comments.

mrfh92 commented on heat/core/io.py:
```diff
@@ -108,6 +112,16 @@ def load_hdf5(
+            if isinstance(load_fraction, float):
+                if load_fraction <= 0.0 or load_fraction >= 1.0:
+                    raise ValueError("load_fraction must be strictly between 0. and 1.")
+                if split is None:
+                    raise ValueError("load_fraction can only be used if split is not None.")
```
I have chosen 2.

mrfh92 commented on heat/preprocessing/preprocessing.py:
```diff
@@ -0,0 +1,612 @@
+        if self.copy:
+            Y = X * self.scale_
+            return Y
+        else:
```
if #1184 has been done, this will make a difference i guess (?)

mrfh92 replied:
sry misunderstood this ... you're right

mrfh92 commented on heat/preprocessing/preprocessing.py:
```diff
@@ -0,0 +1,612 @@
+        if self.with_scaling:
+            _check_n_features(self.scale_, X)
+        if self.copy:
+            Y = X
+            if self.with_centering:
+                Y -= self.center_
+            if self.with_scaling:
+                Y *= self.scale_
+            return Y
```
I have added `.copy()` here. Full functionality of in-place transformations can be assessed only if #1184 has been implemented.

mrfh92 commented:
I have done the requested changes

ClaudiaComito reviewed (approved):
Thanks for the changes @mrfh92 !
